### ISSUE TITLE
Adding a %grammar-kind declaration?

### ISSUE THREAD

**Issue by ratmice** [softdevteam/grmtools#349]
Before I try and come up with a patch, I figured it would be good to discuss this in an issue,
I was considering potentially adding a declaration `%grammar-kind Original(NoAction)`, etc

One of the problems with this is that it is likely that we want to parse the value by just using `Deserialize` on the `YaccGrammarKind`,
this would at least be the *easiest* way. But it brings about a few issues 

1. `cfgrammar` has Optional deserialization support, so if we deserialized that way `%grammar-kind` would only work with serde feature enabled.  Alternately we could just implement this by hand instead of serde?
2. Some declarations depend upon a specific `%grammar-kind`, we may have to move some checks from `parse_declarations` to `ast.complete_and_validate`.

But it could potentially reduce the number of places that `YaccGrammarKind` needs to be specified (build.rs, nimbleparse command line, etc).  So it seemed like it might be worth considering.

**Comment by ltratt** [comment]
Oops, I missed this one entirely! I think I'm broadly in favour of this, though we'd have to define what happens if a user specifies `YaccGrammarKind` *and* the grammar specifices `%grammar-kind`. Probably "specifying `YaccGrammarKind` overrides `%grammar-kind`" is reasonable?

**Comment by ratmice** [comment]
Yeah, that seems like the logical/reasonable choice to me, I'll look into an exploratory patch for it.

**Comment by ratmice** [comment]
So I have also been thinking about this as related to #490 and `RegexOptions` 
We can see an example in [cttests/src/regex_opt.test](https://github.com/softdevteam/grmtools/blob/master/lrpar/cttests/src/regex_opt.test) of this embedded in yaml.

One of the difficulties I ran into with `%grammar-kind` is that it doesn't fit well with the parsing of other directives, it is a new form of directive in that it actually affects other directives, I think that it should probably have the following properties.

1. Because it affects other directives it needs to come first.
2. Because the affect on directives needs to be consistent throughout the file it should be unique.

Things like `RegexOptions` being multi-key brings about a slightly different need, the following is an idea I have been considering including a `%header` directive.

for yacc:
```
%header {
   grammar-kind Original(YaccOriginalActionKind::NoAction)
}
```

for lex:

```
%header {
  RegexOptions octal !dot_matches_new_line posix_escapes
}
```

in @ltratt comment above he mentions
> we'd have to define what happens if a user specifies `YaccGrammarKind` _and_ the grammar specifices `%grammar-kind`. Probably "specifying `YaccGrammarKind` overrides `%grammar-kind`" is reasonable?

One other wrinkle we have is we need to be able to specify that the grammar-kind is to be discovered later.
I think we can tackle both of these issues at once, by adding a new variant for now I'll call this `YaccKind::SelfDescribing`.

The general idea is that for `YaccKind::SelfDescribing` `%header` must exist, and must be the first directive in the file.
In other formats `%header` remains optional, and conflicts are resolved by the precedence as mentioned in previous comments.

I'm not terribly smitten with the actual `%header { ... }` syntax but it was what came to mind.

**Comment by ratmice** [comment]
Slightly funny issue that we need to disallow the `SelfDescribing` variant from being allowed a valid argument to the directive
for whatever syntax we end up with, e.g. none of the below should be considered valid, e.g. it should be one of `Original` , `Grmtools`, `Eco`, or anything *but* `SelfDescribing`.

```
%grammar-kind SelfDescribing
%yacckind SelfDescribing
%header {
   yacckind SelfDescribing
  }
```

**Comment by ltratt** [comment]
@ratmice I agree with pretty much all of this. A few minor comments:

>    Because it affects other directives it needs to come first.

Agreed.

>   Because the affect on directives needs to be consistent throughout the file it should be unique.

I'm not totally sure what this means, though.

I wonder if `%header` should be `%grmtool` to make clear that it contains "grmtools directives" in whatever syntax we think appropriate (which may or may not be Yacc-ish)?

> for whatever syntax we end up with, e.g. none of the below should be considered valid, e.g. it should be one of Original , Grmtools, Eco, or anything but SelfDescribing.

I think you mean "but not SelfDescribing"?

**Comment by ratmice** [comment]
> > Because the affect on directives needs to be consistent throughout the file it should be unique.
> 
> I'm not totally sure what this means, though.

I meant that something like:
```
%grmtools {
 YaccKind GrmTools
 YaccKind Eco
}
```

Originally the thing I was trying to avoid by referring to *unique* is something like the following:

```
%yacckind Original(UserAction)
%actiontype ()
%yacckind Grmtools
```

The thing of note is that `%actiontype` is only a valid directive for `Original(...)`, so the above just doesn't make sense.

> I wonder if `%header` should be `%grmtool` to make clear that it contains "grmtools directives" in whatever syntax we think appropriate (which may or may not be Yacc-ish)?

Seems wise, would we want the same directive for both lex/yacc here?  i.e. use `%grmtools` for both?

> > for whatever syntax we end up with, e.g. none of the below should be considered valid, e.g. it should be one of Original , Grmtools, Eco, or anything but SelfDescribing.
> 
> I think you mean "but not SelfDescribing"?

Yeah, the word I was fumbling for seems to be except.

I was thinking that variant would use the `#[serde(skip)]` and that they %header would just use serde internally for value formatting, but now I see that isn't actually going to work out because serde is an optional feature.  Manual implementation of the values seems the appropriate thing.

**Comment by ltratt** [comment]
> The thing of note is that %actiontype is only a valid directive for Original(...), so the above just doesn't make sense.

Agreed.

> would we want the same directive for both lex/yacc here? i.e. use %grmtools for both?

Good idea -- yes!

> Manual implementation of the values seems the appropriate thing.

Yes, this seems like a good thing, and also allows us to stabilise the syntax and make it human-friendly (to the extent it can be).

**Comment by ratmice** [comment]
Cool, I think then we have a plan in place for every difficulty I had encountered last time I tried to implement this, 
So I think now its just waiting on me to give this a second try, as such assigning myself.

**Comment by ltratt** [comment]
Best of luck -- looking forward to it :)
